### PR TITLE
multiple requirements.txt files

### LINKED ISSUE #77: support for multiple requirements.txt
(Note: this only applies for Docker-based exporters like AWS, and Airflow)

Currently, when generating the Docker image to run tasks, users need to list the dependencies they want to include. For example, a user may have the following:

```txt
# content of requirements.txt
scikit-learn
pandas
```

However, in some cases, users may have a long list of dependencies. This is often the case when using many ML packages that implement different models (to find which model works best). Unfortunately, long environments are sometimes impossible to set up, up if two or more dependencies have conflicts. e.g. scikit-learn requires `numpy==1.2` but xgboost requires `numpy>1.2`. If that happens, users will be unable to create a single docker image.

To fix this, we should allow them to declare multiple `requirements.txt` files and then generate multiple Docker images; note that this does not require multiple Dockerfile since we can re-use the same one, with a different requirements.txt and tag them differently so they effectively become multiple images.

To keep things simple, we can have a naming convention. For example, if a DAG has tasks: load, clean, fit-0, fit-1, and fit-2. We could do something like this:

If a user has `requirements.txt` and nothing else, then generate a single image.

If a user has `requirement.txt`, and any other files that match the format `requirements.*.txt`: match filenames to task names, see below.

## Pattern matching

(we can use [fnmatch](https://docs.python.org/3/library/fnmatch.html) for this)

Say the user has `requirements.txt`, and `requirements.fit-*.txt`

use `requirements.fit-*.txt` for tasks fit-0, fit-1, and fit-2.
use `requirements.txt` for all the other tasks

Open question: is the `*` problematic? I gave it a try and doesn't seem to be any issues on UNIX if a filename has `*` on it. The other thing is to ensure that docker allows `*` in image names. If not, we should use another wildcard, (perhaps double underscore? `__`)

Changes should be applied [here](https://github.com/ploomber/soopervisor/blob/master/src/soopervisor/commons/docker.py) 

The `docker.build` function currently returns a list of image names, with this change, it should extend it to return all the generated images.

## Modifying aws batch exporter

Once images are pushed, the [aws batch exporter](https://github.com/ploomber/soopervisor/blob/master/src/soopervisor/aws/batch.py) needs to pattern match the task names to the right images.

## testing

We should test that given some combinations of  `requirements.txt` files, the `docker` CLI is called with the right arguments and that the tasks are scheduled in AWS with the right arguments as well.

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes

The changes will allow users to declare multiple task specific requirements.*.lock.txt / environment.*.lock.txt files. Multiple images are being built with different dependency files for different task patterns like fit-__, etc. Scripts which do not need separate requirements file will use the image corresponding to the default requirements.lock.txt file. As of now scheduling jobs as per this change will occur only in the AWS batch exporter. 

## Issue ticket number and link
Closes #77 

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have added thorough tests (when necessary).
- [x] I have added the right documentation (when needed). Product update? If yes, write one line about this update.

### CODE REVIEW

**Comment by neelasha23** [comment]
I would need some help here. I'm almost done with the code and test cases. There are some existing test cases which are failing and I cannot get to the root cause even after debugging. 
1. Testcases for airflow export: `test_airflow_export_sample_project` and `test_export_airflow_callables` with the below error

```
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x16ee15310>, mock_docker_calls_callables = <ploomber.io._commander_tester.CommanderTester object at 0x16ee152b0>
tmp_callables = PosixPath('/private/var/folders/ff/jydfr9497jn4lqn28913mvwc0000gn/T/pytest-of-neelashasen/pytest-216/test_export_airflow_callables_2/assets/callables'), no_sys_modules_cache = None
skip_repo_validation = None, mode = 'force', args = ' --force'

    @pytest.mark.parametrize('mode, args', [
        ['incremental', ''],
        ['regular', ''],
        ['force', ' --force'],
    ],
                             ids=['incremental', 'regular', 'force'])
    def test_export_airflow_callables(monkeypatch, mock_docker_calls_callables,
                                      tmp_callables, no_sys_modules_cache,
                                      skip_repo_validation, mode, args):
        exporter = AirflowExporter.new(path_to_config='soopervisor.yaml',
                                       env_name='serve')
    
        # this requires a git repo
        git_init()
    
        exporter.add()
        exporter.export(mode=mode)
    
        monkeypatch.syspath_prepend(Path('serve').resolve())
    
>       mod = importlib.import_module('callables')

/Users/neelashasen/Dev/soopervisor/tests/airflow/test_airflow_export.py:151: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/neelashasen/miniconda3/envs/soopervisor/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:986: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:680: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:850: in exec_module
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
/private/var/folders/ff/jydfr9497jn4lqn28913mvwc0000gn/T/pytest-of-neelashasen/pytest-216/test_export_airflow_callables_2/assets/callables/serve/callables.py:21: in <module>
    spec = json.loads(path_to_spec.read_text())
/Users/neelashasen/miniconda3/envs/soopervisor/lib/python3.9/pathlib.py:1266: in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
/Users/neelashasen/miniconda3/envs/soopervisor/lib/python3.9/pathlib.py:1252: in open
    return io.open(self, mode, buffering, encoding, errors, newline,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = PosixPath('/private/var/folders/ff/jydfr9497jn4lqn28913mvwc0000gn/T/pytest-of-neelashasen/pytest-216/test_export_airflow_callables_2/assets/callables/serve/callables.json')
name = '/private/var/folders/ff/jydfr9497jn4lqn28913mvwc0000gn/T/pytest-of-neelashasen/pytest-216/test_export_airflow_callables_2/assets/callables/serve/callables.json', flags = 16777216, mode = 438

    def _opener(self, name, flags, mode=0o666):
        # A stub for the opener argument to built-in open()
>       return self._accessor.open(self, flags, mode)
E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/ff/jydfr9497jn4lqn28913mvwc0000gn/T/pytest-of-neelashasen/pytest-216/test_export_airflow_callables_2/assets/callables/serve/callables.json'

/Users/neelashasen/miniconda3/envs/soopervisor/lib/python3.9/pathlib.py:1120: FileNotFoundError
```

2. Test cases for argo export :`test_export`, `test_custom_volumes`, `test_export_with_null_repository`

```
mock_docker_my_project = <ploomber.io._commander_tester.CommanderTester object at 0x177b9d460>, backup_packaged_project = None, skip_repo_validation = None
capsys = <_pytest.capture.CaptureFixture object at 0x177b9da90>

    def test_export_with_null_repository(mock_docker_my_project,
                                         backup_packaged_project,
                                         skip_repo_validation, capsys):
        exporter = ArgoWorkflowsExporter.new(path_to_config='soopervisor.yaml',
                                             env_name='serve')
        exporter.add()
    
        # set empty repository
        spec = yaml.safe_load(Path('soopervisor.yaml').read_text())
        spec['serve']['repository'] = None
        Path('soopervisor.yaml').write_text(yaml.safe_dump(spec))
    
        # reload exporter
>       ArgoWorkflowsExporter.load(path_to_config='soopervisor.yaml',
                                   env_name='serve').export(mode='incremental',
                                                            until=None)

/Users/neelashasen/Dev/soopervisor/tests/argo/test_argo_export.py:140: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/neelashasen/Dev/soopervisor/src/soopervisor/abc.py:340: in export
    return self._export(cfg=self._cfg,
/Users/neelashasen/Dev/soopervisor/src/soopervisor/argo/export.py:64: in _export
    pkg_name, target_image = docker.build(cmdr,
/Users/neelashasen/Dev/soopervisor/src/soopervisor/commons/docker.py:193: in build
    image = build_image(e,
/Users/neelashasen/Dev/soopervisor/src/soopervisor/commons/docker.py:95: in build_image
    e.run('docker',
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ploomber.io._commander.Commander object at 0x1612f64f0>, description = 'Testing File client', capture_output = True, expected_output = 'True\n', error_message = 'Missing File client'

    def run(self,
            *cmd,
            description=None,
            capture_output=False,
            expected_output=None,
            error_message=None,
            hint=None,
            show_cmd=True):
        """Execute a command in a subprocess
    
        Parameters
        ----------
        *cmd
            Command to execute
    
        description: str, default=None
            Label to display before executing the command
    
        capture_output: bool, default=False
            Captures output, otherwise prints to standard output and standard
            error
    
        expected_output: str, default=None
            Raises a RuntimeError if the output is different than this value.
            Only valid when capture_output=True
    
        error_message: str, default=None
            Error to display when expected_output does not match. If None,
            a generic message is shown
    
        hint: str, default=None
            An optional string to show when at the end of the error when
            the expected_output does not match. Used to hint the user how
            to fix the problem
    
        show_cmd : bool, default=True
            Whether to display the command next to the description
            (and error message if it fails) or not. Only valid when
            description is not None
        """
        cmd_str = ' '.join(cmd)
    
        if expected_output is not None and not capture_output:
            raise RuntimeError('capture_output must be True when '
                               'expected_output is not None')
    
        if description:
            header = f'{description}: {cmd_str}' if show_cmd else description
            self.tw.sep('=', header, blue=True)
    
        error = None
    
        # py 3.6 compatibility: cannot use subprocess.run directly
        # because the check_output arg was included until version 3.7
        if not capture_output:
            try:
                result = subprocess.check_call(cmd)
            except Exception as e:
                error = e
        # capture outpuut
        else:
            try:
                result = subprocess.check_output(cmd)
            except Exception as e:
                error = e
            else:
>               result = result.decode(sys.stdout.encoding)
E               AttributeError: 'NoneType' object has no attribute 'decode'

/Users/neelashasen/miniconda3/envs/soopervisor/lib/python3.9/site-packages/ploomber/io/_commander.py:132: AttributeError
```

This test probably needs some fix in the conftest.py file but I'm not able to figure out what exactly.

**Comment by neelasha23** [comment]
I also see this error in the Windows CI:
```
Checking out the ref
  "C:\Program Files\Git\bin\git.exe" checkout --progress --force refs/remotes/pull/86/merge
  Error: error: invalid path 'tests/assets/multiple_requirements_project/requirements.clean-*.lock.txt'
  Error: The process 'C:\Program Files\Git\bin\git.exe' failed with exit code 128
```

The repo checkout itself is failing because of this file `tests/assets/multiple_requirements_project/requirements.clean-*.lock.txt`. Probably because we can't have `*` in filename on Windows?

**Comment by idomic** [comment]
@neelasha23 Maybe we can try double underscore as well for windows? __

**Comment by neelasha23** [comment]
There is an issue that I'm facing with the airflow tests. The `tests/airflow/test_airflow_export.py ` cases kept failing because the mock was expecting `pkg-name.json` file inside `env-name` folder for non bash presets (I'm not sure why this happened because I have hardly changed anything in the conftest). So just to see if the test case is passing I changed this line  

https://github.com/ploomber/soopervisor/blob/81d0bda8879980b86c75aafcd4d68facf2130423/src/soopervisor/airflow/export.py#L98

However now the current CI error comes up `FAILED tests/test_export.py::test_checks_the_right_spec_pkg[AirflowExporter] - FileNotFoundError: [Errno 2] No such file or directory: 'serve/my_project.json'` .

I would need some help in understanding this functionality.

**Comment by idomic** [comment]
@neelasha23 this code takes the final Airflow dag and saving it into the file the users can deploy with.
We're opening and writing into the json file, I don't think removing this line will suffice since the whole test is checking this export functionality. @edublancas thoughts?

**Comment by neelasha23** [comment]
Just to clarify, I haven't removed that line. Previously it was writing to pkgname.json I have only modified it to write to envname/ pkgname.json, because that's what the test case is suddenly expecting.  I'm not sure how this happened because I haven't made any major changes in the airflow pipeline.

**Comment by edublancas** [comment]
ok, I found the issue. I pushed a commit changing this line:

https://github.com/ploomber/soopervisor/blob/81d0bda8879980b86c75aafcd4d68facf2130423/src/soopervisor/airflow/export.py#L98

to:

```python
path_dag_dict_out = Path(pkg_name + '.json')
```

since that's what I saw in the master branch.

then I saw a bunch of tests fail.

the problem is here: 

https://github.com/ploomber/soopervisor/blob/81d0bda8879980b86c75aafcd4d68facf2130423/src/soopervisor/commons/docker.py#L245

this line will behave differently depending on what happens with the for loop. If `lock_paths` is empty, it will never execute, if there are 1,2,3 items, it'll execute that number of times. This will cause the function to end up in a different working directory, which is causing the issue. These internal functions should be deterministic, Ideally, they should not change the working directory. and if they do, they should always revert to the original working directory where they were executed

**Comment by neelasha23** [comment]
Got it. I will need to go the parent directory of `dist` before start of every loop otherwise it fails. I will try to think of a way around.
What about the the issue with windows? Doesn't look like filenames can have `*` in them on Windows. So should I change the naming convention of the requirements.task-*.lock.txt files? Maybe have __ like Ido suggested?

**Comment by edublancas** [comment]
yeah, I think double underscore makes sense, we already have something similar for the SLURM exporter: https://github.com/ploomber/soopervisor/blob/566d7773a3af408046d0efbdbd80ad10f77bb620/src/soopervisor/shell/export.py#L33

**Comment by neelasha23** [comment]
The CI is passing now for all the platforms. I have also submitted a pipeline to AWS batch and verified that the tasks are getting scheduled in the corresponding image. Had to use `args = ['docker', 'buildx', 'build', '--platform=linux/amd64', '.', '--tag', image_local]` for the jobs to run successfully.

File: src/soopervisor/commons/dependencies.py [hunk 1]
```diff
@@ -13,3 +71,14 @@ def check_lock_files_exist():
 pip: pip freeze > requirements.lock.txt
 conda: conda env export --no-build --file environment.lock.yml
 """)
+
+    if not _no_missing_dependencies('requirements', 'txt') \
```
Comment (idomic) [inline src/soopervisor/commons/dependencies.py:75]: I think we can simplify this by removing the nots and changing the condition to or

File: tests/aws_batch/test_export.py [hunk 2]
```diff
@@ -165,6 +165,112 @@ def test_export(mock_batch, mock_docker_my_project_serve, monkeypatch,
...
 
+@pytest.mark.parametrize(
+    'mode, args',
+    [
+        ['incremental', []],
+        ['regular', []],
+        ['force', ['--force']],
+    ],
+)
+def test_export_multiple_images(mock_batch, monkeypatch,
```
Comment (idomic) [inline tests/aws_batch/test_export.py:176]: Can we break this test into smaller tests? It'll ease CI debugging when stuff breaks. Also the initial mocks can be a fixture

**Review by idomic — CHANGES_REQUESTED** [review]
Added a few comments, I'll also try now running it and seeing how it works

File: tests/aws_batch/test_export.py (hunk 2 above)
Comment (neelasha23) [reply tests/aws_batch/test_export.py]: Made the changes. If the fixtures look fine will add them to the existing testcases also

**Comment by neelasha23** [comment]
> @neelasha23 I ran it and looked on the code build for this user: 95de7987, with the sample pipeline - multiple_requirements_project. Where can I see the 2 images creation? I didn't see it in the build. Also, I saw the batch tasks are failing (sent you the links).

have sent you some links of job run in us-west-2 (that's where I set up AWS Batch initially). Let me know if I should repeat the same in prod

File: tests/assets/multiple_requirements_project/requirements.lock.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+pandas
+numpy
```
Comment (idomic) [inline tests/assets/multiple_requirements_project/requirements.lock.txt:2]: @neelasha23 Can we make sure those dependencies are completely different between the tasks?

**Comment by neelasha23** [comment]
Have added `seaborn` dependency for the `plot` task in multiple_requirements_project in `test/assets`. While testing it i found a bug. The `docker build` command was installing dependencies from the `env` folder whereas the correct task specific file was present in the `/project` folder.  So i have created a `Dockerfile` specific to aws-batch : https://github.com/ploomber/soopervisor/blob/789f6087a8ec375c79ea97bc8037d8c38636513a/src/soopervisor/assets/aws-batch/Dockerfile and changed the sequence of instalment command to be after `RUN tar --strip-components=1 -zxvf *.tar.gz`.

**Comment by idomic** [comment]
@neelasha23 Is it now batch-specific? Since it wasn't the initial intention.

**Comment by neelasha23** [comment]
No it's not aws batch specific. Since aws batch is the priority right now, I have only modified the Dockerfile  for the same. To be on the safe side I haven't modified the common `Dockerfile`. I did test with Argo and it seemed to be fine, but still once all the exporters are properly tested with this Dockerfile change maybe we can have just one Dockerfile.

**Comment by idomic** [comment]
Got it! Let's finish testing on the rest of the backs, generalize this Dockerfile and merge

**Comment by idomic** [comment]
@edublancas ?
I think it's residues from the run?

**Comment by idomic** [comment]
@neelasha23 how much time do you think it'll take you including the rest of them?
@edublancas thoughts on including the rest of the backends in? If it's ~2-3 days extra dev work I think we should add it, there's no pressing deadline currently.

**Comment by neelasha23** [comment]
I had initially started with Airflow and Kubeflow. And i remember the environment was breaking. So would to figure those out first and then check the code changes. All the exporters have different logic for the export.

**Comment by neelasha23** [comment]
Sure. Any other reviews for this PR?

**Pull request merged**